### PR TITLE
Fixes

### PR DESCRIPTION
**Pull request by rustyrussell**

### CODE REVIEW

File: javascript/index.js [hunk 1]
```diff
@@ -87,23 +91,51 @@ function signature_valid(tlv){
     return merkle_nodes[0]
 }
 // console.log(tlv_offer[8][1](['1000']))
-function decode(paymentRequest){
-    if (typeof paymentRequest !== 'string') throw new Error('Lightning Payment Request must be string')
-    // FIXME: Test vectors
-    paymentRequest=paymentRequest.replace('+','')
-    paymentRequest=paymentRequest.replace(' ','')
+console.log(decode('lno1qcp4256ypqpq86q2pucnq42ngssx2an9wfujqerp0y2pqun4wd68jtn00fkxzcnn9ehhyec6qgqsz83qfwdpl28qqmc78ymlvhmxcsywdk5wrjnj36jryg488qwlrnzyjczlqsp9nyu4phcg6dqhlhzgxagfu7zh3d9re0sqp9ts2yfugvnnm9gxkcnnnkdpa084a6t520h5zhkxsdnghvpukvd43lastpwuh73k29qsy'))
```
Comment (rustyrussell) [inline javascript/index.js]: Should probably remove this debugging line :)

File: javascript/index.js [hunk 2]
```diff
@@ -87,23 +91,51 @@ function signature_valid(tlv){
...
+            i++;
+            while(i<paymentRequest.length && paymentRequest[i]==' '){
+                i++;
+            }
+            e=i;
+            if(e==paymentRequest.length || s==0 || paymentRequest.charAt(s-1) in isBech32==false || paymentRequest.charAt(e) in isBech32==false)
+                throw new Error('Lightning Payment Request must be string')
+            paymentRequest=paymentRequest.slice(0,s)+paymentRequest.slice(e);
+        }
+    }
```
Comment (rustyrussell) [inline javascript/index.js:114]: I think it's simpler to use regular expressions?  Like: `paymentReq.replace(/\+[ \r\n]*/,'')`

File: javascript/index.js [hunk 3]
```diff
@@ -87,23 +91,51 @@ function signature_valid(tlv){
...
+    if(paymentRequest.indexOf(' ')!=-1||paymentRequest.indexOf('+')!=-1)
+        throw new Error('Lightning Payment Request must be string');
```
Comment (rustyrussell) [inline javascript/index.js]: lastIndexOf() is actually correct here...

And Error should be "must contain 1" instead.

File: javascript/index.js [hunk 4]
```diff
@@ -87,23 +91,51 @@ function signature_valid(tlv){
...
+    
+    if (paymentRequest.slice(0, 2) != 'ln') 
+        throw new Error('Not a proper lightning payment request');
 
-    // FIXME: MUST all be UPPER or all LOWER
-    for(let i=0; i<paymentRequest.length; i++){
-        if(!paymentRequest.charAt(i) in isBech32)
-            throw new Error('Not a proper lightning payment request')
+    if (paymentRequest.indexOf('1')==-1)
+        throw new Error('Separator not present');
```
Comment (rustyrussell) [inline javascript/index.js]: Just get the entire human readable prefix, like so:

```
hrplen = payRequest.lastIndexOf('1')
if (hrplen == -1)
     throw new Error('Separator not present')
prefix = payRequest.slice(0,hrplen)
encodeddata = payRequest.slice(hrplen+1)
```

File: javascript/index.js [hunk 5]
```diff
@@ -119,10 +151,8 @@ function decode(paymentRequest){
...
             break;
     default:
-	// FIXME: Mention prefix
-        throw new Error('Not a proper lightning payment request')  
+        throw new Error(prefix.toString() + ' is not a proper lightning prefix')  
     }
-    // console.log(type)
     for (let i=0;i<encodedData.length;i++){
         words[words.length]=ALPHABET_MAP[encodedData.charAt(i)]
     }
```
Comment (rustyrussell) [inline javascript/index.js:154]: Here, not above, is where you should check if a character is in the map?  We're iterating anyway.

File: javascript/index.js [hunk 6]
```diff
@@ -158,7 +189,7 @@ function decode(paymentRequest){
         
 
 	// FIXME: BigSize here too..
-        tagLength = words_8bit.slice(0,1)
+        tagLength = ((words_8bit.slice(0,1)));
```
Comment (rustyrussell) [inline javascript/index.js]: Hmm, what do these (( and )) do?

File: javascript/Makefile [hunk 7]
```diff
@@ -5,7 +5,7 @@ default: gencode.js
 TOPDIR := ..
 
 gencode.js: $(TOPDIR)/specs/bolt12.csv $(TOPDIR)/specs/bolt4.csv $(TOPDIR)/specs/bolt1.csv preamble.js $(TOPDIR)/tools/generate-code.py
-	$(TOPDIR)/tools/generate-code.py --preamble=preamble.js --postamble=postamble.js --language=js > $@
+	python $(TOPDIR)/tools/generate-code.py --preamble=preamble.js --postamble=postamble.js --language=js > $@
```
Comment (rustyrussell) [inline javascript/Makefile:8]: OK, good fix!

File: javascript/index.js [hunk 8]
```diff
@@ -160,58 +159,64 @@ function decode(paymentReq){
...
+    buffer= (Buffer.from(words_8bit))
+    
+    while(buffer.length){
+        let tlvs=[];
+
+        let res=fromwire_bigsize(buffer);
+        
+        const tagCode=res[0];
+        
+        tlvs.push(Number(''+tagCode));
```
Comment (rustyrussell) [inline javascript/index.js:185]: This coersion is now unnecessary I think: fromwire_bigsize() returns an integer.

File: javascript/index.js [hunk 9]
```diff
@@ -160,58 +159,64 @@ function decode(paymentReq){
...
-        // parser = TAGPARSER[tagCode][2]
-	// FIXME: BigSize here too..
-        words_8bit = words_8bit.slice(1)
-        // console.log(TAGPARSER[tagCode])
-        // console.log(tagCode)
-	// FIXME: DUPLICATE!
-        if(tagCode=='0'){
-            break
+            break;        
         }
```
Comment (rustyrussell) [inline javascript/index.js]: Don't break on 0...

File: javascript/index.js [hunk 10]
```diff
@@ -160,58 +159,64 @@ function decode(paymentReq){
...
         
+        res=fromwire_bigsize(buffer);
+        
+        tagLength = res[0];
+        
+        tlvs.push(Number(''+tagLength));
+        
+        buffer = res[1];
+        
+        tagWords = buffer.slice(0, Number(''+tagLength));
```
Comment (rustyrussell) [inline javascript/index.js:197]: Same here

File: javascript/index.js [hunk 11]
```diff
@@ -160,58 +159,64 @@ function decode(paymentReq){
...
+        buffer=buffer.slice(Number(''+tagLength))
+        
+        if(tagCode<240||tagCode>1000)
+            tags.push(Buffer.concat([Buffer.from(tlvs.slice(0,2)),tlvs[2]]).toString('hex'))
     }
-    final['offer_id']=signature_valid(tlv);
     final['contents']=fin_content;
-    return final;
+    final['offer_id']=signature_valid(tags);
+    console.log(final)
```
Comment (rustyrussell) [inline javascript/index.js]: This really needs to be return final not a debug msg :)

File: javascript/index.js (hunk 1 above)
Comment (rustyrussell) [reply javascript/index.js]: Oops, you did later.  Cool!

File: javascript/index.js (hunk 11 above)
Comment (adi2011) [reply javascript/index.js]: Yeah, wrote it while testing locally, removing all the commented and redundant code lines..

File: javascript/index.js (hunk 2 above)
Comment (adi2011) [reply javascript/index.js:114]: Yeah tried this but was not working for some reason so just linear searched the string..

File: javascript/index.js (hunk 3 above)
Comment (adi2011) [reply javascript/index.js]: Umm, this makes sure that after removing all the whitespaces followed by a '+', there should not be any whitespace in the string. But `paymentRequest.indexOf('+')!=-1` should be removed..

File: javascript/index.js (hunk 4 above)
Comment (adi2011) [reply javascript/index.js]: Yes, I did that, this is an older commit, it seems..

**Pull request merged**